### ISSUE TITLE
Training issue

### ISSUE THREAD

**Issue by YitongD** [KevinMathewT/vision-experiments#2]
Hello, I retrained the model following your provided configuration, but I found that the performance is much worse than the released checkpoint best_loss_0.480798_iter_90000_epoch_45.pth.

In addition, during training I noticed that the loss often becomes negative.
Could you please explain what might be the reason for this behavior?

Thank you very much for your help.

**Comment by KevinMathewT** [comment]
Hi @YitongD -- that is actually very interesting, thanks for informing -- so the released weights were trained using the [streaming version](https://github.com/KevinMathewT/dynadust3r-unofficial/blob/main/loaders/stereo4d_stream.py), can you try training using that and seeing if that works? the dataset has to be created using [this script](https://github.com/KevinMathewT/dynadust3r-unofficial/blob/main/extras/preprocess_stereo4d.py). I did confirm that the behaviour of the [direct version](https://github.com/KevinMathewT/dynadust3r-unofficial/blob/main/loaders/stereo4d.py) is the same -- but I'll have another look at it, apologies for this.

Thanks,
Kevin

**Comment by YitongD** [comment]
Thank you very much for your reply.
I will try the streaming version as you suggested.

**Comment by YitongD** [comment]
One more question regarding the loss behavior:
Although theoretically the loss can take negative values, does this indicate that the training might be unstable?

I also noticed that all the checkpoints you provided seem to correspond to positive loss values.
Could you please clarify whether negative losses are expected in practice, or if they suggest an issue in the training setup?

Thank you again for your time and help.

**Comment by KevinMathewT** [comment]
Hi @YitongD -- I was able to find the logs for that run which produced the weights [here](https://huggingface.co/KevinMathew/dynadust3r-unofficial-weights), you can have a look at that for your reference; it might also help clarify some of the queries you have: [stdout.log](https://github.com/user-attachments/files/24292249/stdout.log)

**Comment by YitongD** [comment]
Thank you very much for your reply — it has resolved part of my confusion.

In addition, during training visualization, I noticed that the regions with ground-truth motion (valid) only occupy a very small portion of the image, and they usually do not include moving objects. This might make it difficult to effectively supervise the motion component, which could potentially lead to convergence issues.

I would appreciate your thoughts on whether this observation is expected, and if there are any recommended strategies to address this.

Thanks again for your help.

**Comment by KevinMathewT** [comment]
Hi @YitongD -- apologies for the late reply, I was travelling.

It's possible that the example that got selected for the visualization on wandb had very few regions with ground truth motion -- I would suggest looking at other examples as well, I don't think this is the case in all the videos across the dataset -- however there certainly are some videos which have very few ground truth tracks.

Also as long as there are a decent amount of valid pixels across the entire dataset, the model should be able to learn, since we only supervise on the valid pixel positions; the dataset is decently big, and across the dataset the model should get a good idea of all sorts of cases and 3d interactions and should generalize decently. 
Also there are certain parts of the image where it's difficult to predict a point cloud position at all (for e.g. the sky, or a sea etc.), in these cases the confidence based loss is very helpful, as the model does have the capability to admit that it's difficult to predict a 3d point for a particular pixel, and hence is penalized less if it's incorrect in it's prediction.

Thanks,
Kevin
